### PR TITLE
Add strict mode check to for-in and for-of header in parser_parse_for_statement_start

### PR DESCRIPTION
**Pull request by szilagyiadam**
JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-parser-statm.c [hunk 1]
```diff
@@ -1168,15 +1168,22 @@ parser_parse_for_statement_start (parser_context_t *context_p) /**< context */
...
-
-      /* TODO: remove this after the pre-scanner supports strict mode detection. */
-      if (context_p->next_scanner_info_p->source_p == context_p->source_p
-          && context_p->next_scanner_info_p->type == SCANNER_TYPE_LET_EXPRESSION)
+      if (context_p->status_flags & PARSER_IS_STRICT)
       {
-        scanner_release_next (context_p, sizeof (scanner_info_t));
+        parser_raise_error (context_p, PARSER_ERR_FOR_IN_DECLARATION);
+      }
+      else
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-statm.c]: This else block is not needed.

File: tests/jerry/es2015/forin-header-strict.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+try {
+  eval('for (var i = 0 in {}) {}');
+}
+catch(e) {
```
Comment (rerobika) [inline tests/jerry/es2015/forin-header-strict.js]: Move the catch one line upwards.

File: jerry-core/parser/js/js-parser-util.c [hunk 3]
```diff
@@ -951,6 +951,10 @@ parser_error_to_string (parser_error_t error) /**< error code */
     {
       return "Incorrect use of yield keyword.";
     }
+    case PARSER_ERR_FOR_IN_DECLARATION:
+    {
+      return "for-in loop variable declaration may not have an initializer.";
+    }
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-util.c]: What about for-of? Is this rule ES2015 only?

File: tests/jerry/es2015/forin-header-strict.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+try {
+  eval('for (var i = 0 in {}) {}');
```
Comment (galpeter) [inline tests/jerry/es2015/forin-header-strict.js]: There should be an `assert(false)` after the eval to make sure that the eval did throw an error,

File: jerry-core/parser/js/js-parser-util.c [hunk 5]
```diff
@@ -951,6 +951,14 @@ parser_error_to_string (parser_error_t error) /**< error code */
     {
       return "Incorrect use of yield keyword.";
     }
+    case PARSER_ERR_FOR_IN_DECLARATION:
+    {
+      return "for-in loop variable declaration may not have an initializer.";
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-util.c]: I`d use only one new error message like `for in-of loop ...`

**Pull request merged**